src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -110,27 +126,122 @@ impl WorkUnitStore {
     }))
   }
 
-  pub fn get_workunits(&self) -> Arc<Mutex<WorkUnitInnerStore>> {
-    self.inner.clone()
+  pub fn get_workunits(&self) -> Vec<WorkUnit> {
+    let mut inner_guard = (*self.inner).lock();
+    let inner_store: &mut WorkUnitInnerStore = &mut *inner_guard;
+    let workunit_records = &inner_store.workunit_records;
+    inner_store
+      .completed_ids
+      .iter()
+      .flat_map(|id| workunit_records.get(id))
+      .flat_map(|record| match record {
+        WorkunitRecord::Started(_) => None,
+        WorkunitRecord::Completed(c) => Some(c.clone()),
+      })
+      .collect()
+  }
+
+  pub fn start_workunit(
+    &self,
+    span_id: SpanId,
+    name: String,
+    parent_id: Option<SpanId>,
+    metadata: WorkunitMetadata,
+  ) -> SpanId {
```

Comment:
To make the `start`/`complete` pattern a bit easier, one approach is to wrap the return value in a newtype, and then give it a `impl Drop` that calls complete.

The caller would probably still want to call `mem::drop` explicitly somewhere. But it avoids double-drops, or illegal tokens.

Reply:
That's not a bad idea, but I think I'd rather make it a separate PR rather than try to get it merged in this one.